livebooks/planned-disruptions.livemd

```diff
@@ -0,0 +1,173 @@
+# Planned Disruptions
+
+## Section
+
+```elixir
+defmodule Dotcom.Utils.Datetime do
+  def in_range?({start, nil}, datetime) do
+    Timex.after?(datetime, start)
+  end
+
+  def in_range?({start, stop}, datetime) do
+    Timex.after?(datetime, start) && Timex.before?(datetime, stop)
+  end
+
+  def later() do
+    {_, end_of_next_week} = next_week()
+    beginning_of_later = Timex.shift(end_of_next_week, seconds: 1)
+
+    {beginning_of_later, nil}
+  end
+
+  def later?(datetime) do
+    later()
+    |> in_range?(datetime)
+  end
+
+  def next_week() do
+    now = Timex.now()
+    end_of_week = Timex.end_of_week(now) |> Util.end_of_service()
+    beginning_of_next_week = Timex.shift(end_of_week, seconds: 1)
+    end_of_next_week = beginning_of_next_week |> Timex.end_of_week() |> Util.end_of_service()
+
+    {beginning_of_next_week, end_of_next_week}
+  end
+
+  def next_week?(datetime) do
+    next_week()
+    |> in_range?(datetime)
+  end
+
+  def overlap?(range, {start, stop}) do
+    in_range?(range, start) || in_range?(range, stop)
+  end
+
+  def this_week() do
+    now = Timex.now()
+    end_of_week = now |> Timex.end_of_week() |> Util.end_of_service()
+
+    {now, end_of_week}
+  end
+
+  def this_week?(datetime) do
+    this_week()
+    |> in_range?(datetime)
+  end
+
+  def today() do
+    now = Timex.now()
+    end_of_day = now |> Timex.end_of_day() |> Util.end_of_service()
+
+    {now, end_of_day}
+  end
+
+  def today?(datetime) do
+    today()
+    |> in_range?(datetime)
+  end
+end
+```
+
+```elixir
+defmodule PlannedDisruptions.Logic do
+  @moduledoc false
+
+  import Dotcom.SystemStatus.Alerts, only: [filter_relevant: 1]
+
+  alias Dotcom.Utils.Datetime
+
+  @buckets %{
+    later: [],
+    next_week: [],
+    this_week: [],
+    today: []
+  }
+  @lines ["Blue", "Green", "Orange", "Red"]
+
+  def lines, do: @lines
+
+  def buckets() do
+    @lines
+    |> Alerts.Repo.by_route_ids(Timex.now())
+    |> filter_relevant()
+    |> Enum.reduce(@buckets, fn alert, buckets ->
+      buckets
+      |> later(alert)
+      |> next_week(alert)
+      |> this_week(alert)
+      |> today(alert)
+    end)
+  end
+
+  for {bucket, _} <- @buckets do
+    defp unquote(bucket)(buckets, alert) do
+      if Enum.any?(alert.active_period, fn range ->
+           Datetime.overlap?(Kernel.apply(Datetime, unquote(bucket), []), range)
```

Comment:
fancy!

Reply:
That's the beauty of having a bunch of small, reusable functions is that you can take on really generic cases without specificity. I could probably figure out a way to do something like this in the pipe of function calls above too.